### DISCUSSION TITLE
PV Output

### DISCUSSION THREAD

**Discussion by TonyM1958** [TonyM1958/HA-FoxESS-Modbus#11] [General]
A discussion area about pvoutput.org including registration, data uploading, maps and teams.

I will do some investigation into the data required and post information here. If you are also looking or want to contribute information, please add to the discussion.

**Comment by TonyM1958** [comment]
**Team Setup**

I've uploaded my data for the last few months and created a system around this.

I've joined the sytem to the team **Fox ESS Owners Group**. To do this, once you have registered, setup a system and uploaded 10 days of output data, go to Settings and find your system (at the bottom of the page) and click Edit. Scroll down to Advanced Settings and find Team Membership. Enter 'Fox ESS Owners Group' in the box and click Join.

At present, I can't see the group but I think that might be because there is a minimum of 10 members before a group is visible.

**Comment by TonyM1958** [comment]
**Output Data**

I've done some digging around on the site and in the forum and there isn't a formal definition of what the Generated value represents e.g. is this the daily PV energy generated or the inverter Output energy - so it's really up to us to decide which number we want to upload.

The only constraint I did find when uploading data is that your export cannot exceed the generated value.

The minimum data required is a Date and Generated energy for that day. I recommend adding your export (feed in) as well. Any other data you want to upload is up to you.

Where you have HA access, the final values for the sensors _pv_energy_today_ and _feed_in_energy_today_ are recommended. If you don't have these, use _pv_energy_daily_ and _feedin_daily_ although I found, if HA has been rebooted during a day, these sensors fall behind the inverter energy meters and under report as a result.

**Reply by Daibutt** [reply]
Managed to delete my old redundant account and set up a new one. Bizarrely you can't delete old systems (so my junk one I added is still in there!) .
Bulk uploaded history data but looks like I messed it up. I did (PV) generation, export and consumption..... but it looks like their definition of consumption may NOT be grid import which is what I uploaded! 
Will have to check and if wrong, delete and re-enter!
Couldn't see FoxESS team yet either.

**Reply by TonyM1958** [reply]
Looking at what the site does, it confuses some of the graphs around Net (generation) and Efficiency when you upload Consumption data because of the battery, There are days when your Consumption and Export are more than your Generation so your Efficiency and Net go negative...

I'd just concentrate on uploading the PV Generation and Feed In data?

**Comment by TonyM1958** [comment]
Yeh, consumption is your load power, grid consumption is called import but I could not see a clean way to upload that. It will let you just over-write existing data but it did prompt me at one point saying wait 5 minutes between CSV uploads so there may also be some rate limiting.

For the teams - go to Settings, find your system and click Edit. Go down to Advanced Settings and find Team Membership. Enter 'Fox ESS Owners Group' and click join. That worked for me. If I do it again, it tells me I already joined that team.

It doesn't show up under Teams yet but I think it need 10 members minimum.

I've got some python scripts to download the Fox cloud raw data and integrate the PV power values to give a reasonable approximation for PV energy today. I've used this to upload data for the days between when my system was installed and when I setup HA. It's not ready for other people to use yet, but it may provide a way to upload data for people that dont have HA.

**Reply by Daibutt** [reply]
Yes after my last post I did just try manually enetering team name and clicking "join" repeatedly, eventually got the "already joined" message, so fingers crossed. 
Edited: will just update PV gen and export after reading your post about net use and batteries etc.
Yes it looks like it takes 200 rows max and I had to wait 5mins between each batch upload.

**Reply by TonyM1958** [reply]
check the units before generating the data - I can set kwh for quite a few inputs but there is no box for Import Off Peak for me, so I guess this means Wh and not kWh?

**Reply by Daibutt** [reply]
If you change import peak to kWh, import off peak also changes it seems (showing sample data button helped me out with this). 
![image](https://github.com/TonyM1958/HA-FoxESS-Modbus/assets/69276073/411cb722-9f74-4a12-8927-744ad749d531)

**Reply by TonyM1958** [reply]
ah ok - got that now - thanks.

**Comment by Daibutt** [comment]
Team is now showing 👍 
![image](https://github.com/TonyM1958/HA-FoxESS-Modbus/assets/69276073/5b84f43d-ed61-4d28-8e55-ca28984e8480)

**Comment by TonyM1958** [comment]
**Historic Data Upload**

For anyone familiar with Python and Jupyter Lab and wanting to upload historic data, there is some code here that generates a CSV upload for a given start date and number of days. The data contains the PV generation, Feed In, Load and Grid Consumption.

https://github.com/TonyM1958/FoxESS-Cloud/blob/main/README.md#pv-output

I used this to populate pvoutput with data for my system prior to setting up Home Assistant. Note: uploading data from Home Assistant, where available, is recommended as this data is more accurate than the Fox Cloud data. There are also some gaps in the Fox Cloud data when data may not have been collected when the service was taken down for maintenance around the middle of April 2023.

**Comment by TonyM1958** [comment]
**HA Data Upload via SQL**

If you have MySQL access to Home Assistant, you can run a query against your statistics data to upload data. The query and details of the settings are here:

https://github.com/TonyM1958/HA-FoxESS-Modbus/wiki/Exporting-Data#pv-output

I used this to populate pvoutput with my data from the day I setup Home Assistant. This takes data directly from the inverter energy meters via the _today sensors available over RS-485. Depending on your integration, you daily energy sensors might have different names.

Also be aware, if you have a second solar inverter, double check the load energy reported by HA as it may not be correct.

**Comment by TonyM1958** [comment]
**Automatic upload from HA**

I now have an automation that uploads energy data to pvoutput every hour. You could just run this once a day but its running more often just now for testing. I'll run it for a day or two to make sure it works before posting details of how to setup it up. The gist of it is this:

+ Add rest_command to configuration.yaml with the rest command for pvoutput_upload
+ Add your api_key and system_id to secrets.yaml
+ Add a template that grabs and formats the date now and energy data for pv, feedin, load and grid consumption today
+ Add an automation that runs every hour and triggers the rest command

**Comment by TonyM1958** [comment]
Think I've worked our a way to take PV generation data from the Fox Cloud API and use this to populate pvoutput. The data I'm using is 5 minute samples so it may even be possible to partition import and export data in Flux peak / off peak periods. May take a few days to work out.

Should be able to run on a PC with Anaconda / Jupyter Lab. Thoughts?

**Reply by Chrisbrown01942** [reply]
Sounds great! Wish I could offer some input, but all this is way above my pay grade. I use InfluxDB to store the data HA collects, so I'm using that to output a CSV file & manually uploading it to PVOutput. Having the process automated will be a big improvement!

**Reply by Daibutt** [reply]
Ditto. Anaconda/Jupyter are not on my radar. 
I do record daily imp/exp/PV/load so have used the CV loader to upload a year + a bit of daily totals. 
Was thinking of maybe doing the same with HA (i.e at 23:58/:59 ish just zap off the current day totals using the API. If that is even possible!)

**Reply by TonyM1958** [reply]
Details of how to automate updates from HA are here: https://github.com/TonyM1958/HA-FoxESS-Modbus/wiki/PV-Output#daily-updates

I've had this running a couple of days and William installed it yesterday. We had a problem setting up the template as he wasn't using the new yaml config structure. If you try it and find any problems, let me know. By default, it updates every hour so you can check its working through the day but you can change the trigger pretty easily.

**Reply by TonyM1958** [reply]
BTW Anaconda and Jupyter Lab are just a way to setup a python run time environment and provide a shell for running stuff. Nowhere near as scary as it sounds but a lot more flexible for refining functional code than developing  a graphical interface when you don't yet know the application dynamics.

Some of the recent additions to EnergyStats, for example, build out from some of the underlying api manipulation code I've created this way... not that I want to steal any thunder from Al - he really makes stuff fly.

**Comment by MrEcoJones** [comment]
Just peeking in to say I'm following this but not sure if I can add much on the development side. 

It is looking interesting so far.

**Comment by Daibutt** [comment]
Brilliant Tony! 
All makes sense to me, and apart from renaming some of the sensors (I'm still on Stealth Chestnut v0.1 and some earlier stuff!)  I'll give this a go tomorrow.

**Reply by Daibutt** [reply]
The more I poke around in PVOutput the more I discover! 
I am quite impressed with a lot of the charting but admit I'm not sure what all of it does yet.
I can see Tony has loaded up 2023 daily data and so have I, so you can compare and trend different systems super easily!
![image](https://github.com/TonyM1958/HA-FoxESS-Modbus/assets/69276073/94105d19-02fb-431c-b8a5-529558a4faae)

**Comment by Daibutt** [comment]
@TonyM1958 Great instructions Tony 👍 , one minor point to maybe add to instructions :
After adding the code etc. I reloaded just templates yaml and when running the automation I got an error. So then reloaded all yaml, ditto. 
I had to do a full HA restart for it to recognise the rest_command service. 
I got data in PVOutput.org immediately, need to check "Import" in a while as it's showing "-" like there's no data rather than 0.0000
![image](https://github.com/TonyM1958/HA-FoxESS-Modbus/assets/69276073/4267e1e3-b4ab-4e2b-833e-b49f64db2cab)

**Reply by Daibutt** [reply]
I did a quick force charge just to get some import and it's all good!
![image](https://github.com/TonyM1958/HA-FoxESS-Modbus/assets/69276073/d582a257-a5c5-4523-a9e5-12ce21ed18a7)

Fantastic work Tony, thank you! 👌

**Reply by TonyM1958** [reply]
glad it worked for you. Yes, I probably restarted HA a few times whilst getting it to work - good tip, I'll add that to the wiki. Any zero value, pvoutput shows as -. Must just be luck that you were up at 2am so you could test force charging at the cheapest rate!

**Reply by TonyM1958** [reply]
BTW, as a note, the automation currently uploads all import as off peak and zeros the other import registers. It sums across the registers so you incorrect totals if you don't do this. Just in case you are also looking at the tariffs and credit / debit it shows. These are calculated at upload so far as I can tell. Recalculation if you update your tariff requires donation to unlock by the look of it.

**Comment by Chrisbrown01942** [comment]
Brilliant Tony, got that working easily. As you probably have guessed I don't find all this HA easy, but your instructions were clear & easy to follow for someone who has been working with HA for a while , but with no previous experience of it or coding. Yay, no more uploading CSV files manually. Thanks!

**Comment by TonyM1958** [comment]
@Chrisbrown01942 @Daibutt @MrEcoJones here's a new trick that I'm looking for people to try out...

We now have the technology to get data from your Fox Cloud account and upload this to PV Output. THIS DOES NOT REQUIRE HA. Previously, I said you might need to install Jupyter Lab for this, but I now have it running in the cloud using Google Colaboratory. This means you just need a Google account with access to Google Drive.

The details of how to run this are here: https://github.com/TonyM1958/FoxESS-Cloud/wiki/Running-in-Google-Colaboratory

What you get is fully scripted access to your Fox Cloud data and settings via Python - if you want to start exploring data etc.

But, what I've given you to start with are some simple cells that generate CSV data from when your inverter was installed that you can use to upload to PV Output. This performs a Riemann Sum on your power data to estimate your true PV energy generation (not the Yield that Fox provides) so you may find the numbers vary slightly from what you expect. In testing, I have found that the results are within 3% of the energy meters built into the inverter. 

In addition, there is a script that, by default, generates your data for yesterday and directly uploads this.

BTW - the scripts also apply time of use to your data (TOU) so you will see your grid import and export energy split out into the Flux peak, off-peak and standard times. If you are on a different tarrif, it is possible to alter the time periods. Here's an example of my data with TOU:

![image](https://github.com/TonyM1958/HA-FoxESS-Modbus/assets/63789168/cc12b252-2629-4a32-8e54-e298014bcd42)



Any volunteers to give it a go and let me know how you get on?

**Reply by Joshuaharry120** [reply]
See I'd love to the the tariff on PVOutput, but as am on IOG, it can be any time it's 7pence, otherwise it's 30pence, with 15pence export(anytime) - Anyway, You can feed within the PVOutput that it's peak/off-peak rate

**Reply by TonyM1958** [reply]
TOU is not available if you are uploading data from HA as it just uploads data from the daily counters and there is no TOU information.

It is available when you upload data using the Fox ESS Cloud API. This is covered here: https://github.com/TonyM1958/FoxESS-Cloud/wiki/Using-Jupyter-Notebook#time-of-use
